README.md
# Investec API library for Clojure

Clojure wrapper for [Investec's Programmable Banking API](https://developer.investec.com/programmable-banking/).

## Installation

[![Clojars Project](https://img.shields.io/clojars/v/net.clojars.walterl/investec-api.svg)](https://clojars.org/net.clojars.walterl/investec-api)

Add `net.clojars.walterl/investec-api {:mvn/version "0.1.1"}` to your dependencies in `deps.edn`.

## Example

This library is intended for use by other Clojure applications.

```clojure
(require '[walterl.investec-api.accounts :as inv])
(require '[walterl.investec-api.auth :as auth])

(def api-token (auth/access-token "<MY CLIENT ID>" "<MY SECRET>"))

(doseq [{:keys [account-id account-name account-number product-name]} (inv/accounts api-token)
        :let [{:keys [currency current-balance]} (inv/balance api-token account-id)]]
  (println (format "Account %s (%s %s) has %s %.2f"
                   account-name product-name account-number
                   currency (float current-balance))))
```

The above example will print the following for the example data from the API docs:

```
Account Mr John Doe (Private Bank Account 10010206147) has ZAR 28857.76
```

See the [API docs](https://walterl.github.io/investec-api) for more details about available functions.

## Progress

* [X] [Authorization](https://developer.investec.com/programmable-banking/#open-api-authorization)
* [X] [Accounts](https://developer.investec.com/programmable-banking/#open-api-accounts)
* [ ] [Programmable cards API](https://developer.investec.com/programmable-banking/#open-api-programmable-cards-api)

## Development

Run the project's tests:

    $ clj -T:build test

Run the project's CI pipeline and build a JAR:

    $ clj -T:build ci

Install it locally (requires the `ci` task be run first):

    $ clj -T:build install

Generate API docs (outputs to `target/doc`):

    $ clj -X:codox

## License

Copyright © 2022 @walterl

Distributed under the Eclipse Public License version 1.0.


CHANGELOG.md
# Change Log
All notable changes to this project will be documented in this file. This change log follows the conventions of [keepachangelog.com](http://keepachangelog.com/).

## 0.1.1 - 2022-03-06
### Added
- First release.
- Implemented _Authorization_ and _Accounts_ sections of [Investec's Programmable Banking API](https://developer.investec.com/programmable-banking/#programmable-banking).


test/walterl/investec_api_test.clj
(ns walterl.investec-api-test
  (:require [clojure.test :refer :all]
            [walterl.investec-api :refer :all]))

(deftest a-test
  (testing "FIXME, I fail."
    (is (= 0 1))))


test/walterl/investec_api/auth_test.clj
(ns walterl.investec-api.auth-test
  (:require [cheshire.core :as json]
            [clojure.test :refer [deftest is testing]]
            [java-time :as t]
            [walterl.investec-api.auth :as auth]
            [walterl.investec-api.http :as http]))

(deftest request-access-token!-test
  (testing "access token"
    (let [client-id "client id"
          secret "hunter2"
          header-date (t/offset-date-time 2022 2 17 0 0 0)
          header-date-str (t/format (t/formatter :rfc-1123-date-time) header-date)
          fake-http-post (fn [url opts]
                           {:body (json/generate-string {:url url, :opts opts})
                            :headers {"date" header-date-str}
                            :status 200})]
      (binding [http/*http-post* fake-http-post]
        (let [resp (#'auth/request-access-token! client-id secret)]
          (testing "response is well formed"
            (is (contains? resp :body))
            (is (contains? resp :headers))
            (is (= header-date-str (get-in resp [:headers "date"])))))))))

(deftest access-token-test
  (testing "auth/access-token"
    (let [client-id "client id"
          secret "l3tmein"
          header-date (t/offset-date-time 2022 2 17 0 0 0)
          header-date-str (t/format (t/formatter :rfc-1123-date-time) header-date)
          access-token "ACC3SST0K3N"
          token-type "Bearer"
          expires-in 1799
          scope "scope"
          post-args (atom [])]
      (binding [http/*http-post*
                (fn [url opts]
                  (swap! post-args conj {:url url, :opts opts})
                  {:body (json/generate-string {:access_token access-token
                                                :expires_in expires-in
                                                :scope scope
                                                :token_type token-type})
                   :headers {"date" header-date-str}
                   :status 200})]
        (let [resp (auth/access-token client-id secret)
              args (first @post-args)]
          (testing "returns access token map"
            (is (= {:access-token access-token
                    :token-type token-type
                    :expires-at (t/plus header-date (t/seconds expires-in))
                    :scope scope}
                   resp)))
          (testing "is requested with correct parameters"
            (is (= (#'http/url "/identity/v2/oauth2/token")
                   (:url args)))
            (is (= (#'auth/basic-auth-header client-id secret)
                   (get-in args [:opts :headers])))
            (is (= "grant_type=client_credentials&scope=accounts"
                   (get-in args [:opts :body])))))))))


test/walterl/investec_api/http_test.clj
(ns walterl.investec-api.http-test
  (:require [cheshire.core :as json]
            [clojure.test :refer [deftest is testing]]
            [walterl.investec-api.http :as http])
  (:import clojure.lang.ExceptionInfo))

(deftest GET-test
  (testing "GET"
    (let [get-args (atom [])
          fake-http-get (fn [url opts]
                          (swap! get-args conj {:url url, :opts opts})
                          {:body (json/generate-string {:url url})
                           :status 200})]
      (binding [http/*http-get* fake-http-get]
        (let [token {:access-token "ACC3SST0K3N" ; these are the only required fields
                     :token-type "Bearer"}
              query-opts {:x-test-opt "This is a test value"}
              resp (http/GET token "path" query-opts)
              req-opts (:opts (last @get-args))]
          (testing "forwards query opts"
            (is (= query-opts
                   (select-keys req-opts (keys query-opts)))))
          (testing "correctly encodes token auth header"
            (is (= (str (:token-type token) " " (:access-token token))
                   (get-in req-opts [:headers "Authorization"]))
                (prn-str (:headers req-opts))))
          (testing "response includes headers, body and status"
            (is (map? resp))
            (is (contains? resp :headers))
            (is (contains? resp :body))
            (is (contains? resp :status)))
          (testing "parses response JSON"
            (is (map? (:body resp)))
            (is (= #{:url}
                   (set (keys (:body resp))))))))
      (binding [http/*http-get* (fn [_url _opts]
                                  {:error "Test error"
                                   :status 500})]
        (testing "throws on non-200 HTTP status"
          (is (thrown-with-msg? ExceptionInfo #"^API request failed$"
                                (http/GET nil ""))))))))


src/walterl/investec_api/http.clj
(ns walterl.investec-api.http
  (:require [cheshire.core :as json]
            [clj-http.lite.client :as http]
            [clojure.string :as str]))

(def ^:private root "https://openapi.investec.com")

(def ^:private ^:dynamic *http-get* http/get)
(def ^:private ^:dynamic *http-post* http/post)

(defn- url
  "Returns API URL for `path`. I.e. it prefixes `root`."
  [path]
  (str root "/" (str/replace-first path #"^/" "")))

(defn- token-auth-header
  [{:keys [access-token token-type]}]
  {"Authorization" (str token-type " " access-token)})

(defn- has-auth-header?
  [opts]
  (some? (get-in opts [:headers "Authorization"])))

(defn- redacted-auth-header
  [opts]
  (update opts :headers assoc "Authorization" "**REDACTED**"))

(defn- safe-query-opts
  [opts]
  (cond-> opts
    (has-auth-header? opts) (redacted-auth-header)))

(defn- request
  [http-method token path query-opts]
  (let [query-opts' (merge query-opts (when token {:headers (token-auth-header token)}))
        {:keys [body headers status] :as resp} (http-method (url path) query-opts')]
    (when (not= 200 status)
      (throw (ex-info "API request failed"
                      {:resp resp
                       :path path
                       :http-method http-method
                       :query-opts (safe-query-opts query-opts')})))
    {:body (json/parse-string body keyword)
     :headers headers
     :status status}))

(defn GET
  "Performs GET request against API, returning parsed and keywordized JSON
  `:body`, `:headers` and `:status`.

  `token` as returned by [[walterl.investec-api.auth/access-token]]."
  [token path & [query-opts]]
  (request *http-get* token path query-opts))

(defn POST
  "Performs POST request against API, returning parsed and keywordized JSON
  `:body`, `:headers` and `:status`.

  `token` as returned by [[walterl.investec-api.auth/access-token]]. May be
  `nil`."
  [token path body & [query-opts]]
  (request *http-post* token path (merge query-opts {:body body})))


src/walterl/investec_api/accounts.clj
(ns walterl.investec-api.accounts
  (:require [camel-snake-kebab.core :as csk]
            [clojure.set :refer [rename-keys]]
            [walterl.investec-api.http :as http]
            [walterl.investec-api.util :as u]))

;;; Get Accounts

(defn- json->account-map
  [m]
  (rename-keys m
               {:accountId :account-id
                :accountName :account-name
                :accountNumber :account-number
                :referenceName :reference-name
                :productName :product-name}))

(defn accounts
  "Obtain a list of accounts in associated profile.

  `token` as returned by [[walterl.investec-api.auth/access-token]].

  https://developer.investec.com/programmable-banking/#get-accounts

      => (accounts token)
      [{:account-id \"172878438321553632224\",
        :account-name \"Mr John Doe\",
        :account-number \"10010206147\",
        :reference-name \"My Investec Private Bank Account\",
        :product-name \"Private Bank Account\"}]
  "
  [token]
  (-> (http/GET token "/za/pb/v1/accounts")
      (get-in [:body :data :accounts])
      (->> (mapv json->account-map))))

;;; Get Account Balance

(defn- json->balance-map
  [m]
  (rename-keys m
               {:accountId :account-id
                :availableBalance :available-balance
                :currentBalance :current-balance}))

(defn balance
  "Obtain a specified account's balance.

  `token` as returned by [[walterl.investec-api.auth/access-token]].

  https://developer.investec.com/programmable-banking/#get-account-balance

      => (balance token \"172878438321553632224\")
      {:account-id \"172878438321553632224\",
       :current-balance 28857.76,
       :available-balance 98857.76,
       :currency \"ZAR\"}
  "
  [token account-id]
  (-> (http/GET token (str "/za/pb/v1/accounts/" account-id "/balance"))
      (get-in [:body :data])
      json->balance-map))

;;; Get Account Transactions

(defn- txn-opts->query-params
  [{:keys [from-date to-date transaction-type]}]
  (cond-> {}
    from-date (assoc "fromDate" (u/iso8601-date-time from-date))
    to-date (assoc "toDate" (u/iso8601-date-time to-date))
    transaction-type (assoc "transactionType" (some-> transaction-type csk/->PascalCaseString))))

(defn- txn-with-parsed-values
  [m]
  (-> m
      (update :card-number not-empty)
      (update :status #(some-> % csk/->kebab-case-keyword))
      (update :type #(some-> % csk/->kebab-case-keyword))
      (update :transaction-type #(some-> % csk/->kebab-case-keyword))))

(defn- json->txn-map
  [m]
  (-> m
      (rename-keys {:accountId :account-id
                    :actionDate :action-date
                    :cardNumber :card-number
                    :postedOrder :posted-order
                    :postingDate :posting-date
                    :runningBalance :running-balance
                    :transactionDate :transaction-date
                    :transactionType :transaction-type
                    :valueDate :value-date})
      txn-with-parsed-values))

(defn transactions
  "Retrieves all transactions for specified account, optionally filtered by
  `:transaction-type`, `:from-date` and/or `:to-date`.

  `token` as returned by [[walterl.investec-api.auth/access-token]].

  https://developer.investec.com/programmable-banking/#get-account-transactions

      => (transactions token
                       \"172878438321553632224\"
                       {:transaction-type :fees-and-interest
                        :from-date (t/minus (t/local-date-time) (t/months 2))
                        :to-date (t/minus (t/local-date-time) (t/months 1))})
      [{:description \"MONTHLY SERVICE CHARGE\",
        :amount 535.0,
        :running-balance 28857.76,
        :card-number nil,
        :posting-date \"2020-06-11\",
        :type :debit,
        :account-id \"172878438321553632224\",
        :status :posted,
        :action-date \"2020-11-10\",
        :transaction-date \"2020-06-10\",
        :transaction-type :fees-and-interest,
        :value-date \"2020-06-10\",
        :posted-order 13379}
       {:description \"CREDIT INTEREST\",
        :amount 31.09,
        :running-balance 29392.76,
        :card-number nil,
        :posting-date \"2020-06-11\",
        :type :credit,
        :account-id \"172878438321553632224\",
        :status :posted,
        :action-date \"2020-11-10\",
        :transaction-date \"2020-06-10\",
        :transaction-type :fees-and-interest,
        :value-date \"2020-06-10\",
        :posted-order 13378}]
  "
  [token account-id opts]
  (-> (http/GET token
               (str "/za/pb/v1/accounts/" account-id "/transactions")
               {:query-params (txn-opts->query-params opts)})
      (get-in [:body :data :transactions])
      (->> (mapv json->txn-map))))


src/walterl/investec_api/auth.clj
(ns walterl.investec-api.auth
  (:require [java-time :as t]
            [walterl.investec-api.http :as http]
            [walterl.investec-api.util :as u]))

(defn- basic-auth-header
  [client-id secret]
  {"Authorization" (str "Basic " (u/base64-encode (str client-id ":" secret)))})

(defn- request-access-token!
  [client-id secret]
  (http/POST nil
             "/identity/v2/oauth2/token"
             "grant_type=client_credentials&scope=accounts"
             {:headers (basic-auth-header client-id secret)}))

(defn- rfc1123->offset-date-time
  [s]
  (t/offset-date-time (t/formatter :rfc-1123-date-time) s))

(defn- safe-rfc1123->offset-date-time
  "Parses date time from RFC1123-formatted `s`tring, or returns \"now\" if it fails."
  [s]
  (or (try
        (rfc1123->offset-date-time s)
        (catch Throwable _))
      (t/offset-date-time)))

(defn- response->access-token
  [{:keys [body headers]}]
  (let [{:keys [access_token expires_in scope token_type]} body]
    {:access-token access_token
     :token-type token_type
     :expires-at (t/plus (safe-rfc1123->offset-date-time (get headers "date"))
                         (t/seconds expires_in))
     :scope scope}))

(defn access-token
  "Obtain an access token.

  https://developer.investec.com/programmable-banking/#get-access-token

      => (access-token \"client-id\" \"secret\")
      {:access-token \"Ms9OsZkyrhBZd5yQJgfEtiDy4t2c\"
       :token-type \"Bearer\"
       :expires-in #object[java.time.OffsetDateTime 0x4304ab94 \"2022-02-17T04:17:54.662519+02:00\"]
       :scope \"accounts\"}
  "
  [client-id secret]
  (response->access-token (request-access-token! client-id secret)))
